src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -168,13 +169,89 @@ private void OnEventSourceChanged(object? _1, TaggerEventArgs _2)
+                new TagSourceQueueItem(highPriority, frozenPartialSemantics, nonFrozenComputationToken),
+                _dataSource.CancelOnNewWork);
+        }
+
+        private async ValueTask ProcessEventChangeAsync(
+            ImmutableSegmentedList<TagSourceQueueItem> changes, CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfTrue(changes.IsEmpty);
+
             // If any of the requests was high priority, then compute at that speed.
-            var highPriority = changes.Contains(true);
-            return new ValueTask<VoidResult>(RecomputeTagsAsync(highPriority, cancellationToken));
+            var highPriority = changes.Contains(x => x.HighPriority);
+
+            // If any of the requests are for frozen partial, then we do compute with frozen partial semantics.  We
+            // always want these "fast but inaccurate" passes to happen first.  That pass will then enqueue the work
+            // to do the slow-but-accurate pass.
+            var frozenPartialSemantics = changes.Contains(t => t.FrozenPartialSemantics);
+
+            if (frozenPartialSemantics)
+            {
+                // If we were asking for frozen partial semantics, then just proceed as normal, getting those tags
+                // quickly, but inaccurately.
+                await RecomputeTagsAsync(highPriority, frozenPartialSemantics: true, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                if (!_dataSource.SupportsFrozenPartialSemantics)
+                {
+                    // We're asking for normal expensive full tags, and this tagger doesn't support frozen partial
+                    // tagging anyways, so just proceed as normal, asking for expensive tags.
+                    await RecomputeTagsAsync(highPriority, frozenPartialSemantics: false, cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    // We're asking for expensive tags, and this tagger supports frozen partial tags.  Kick off the work
+                    // to do this expensive tagging, but attach ourselves to the requested cancellation token so this
+                    // expensive work can be canceled if new requests for frozen partial work come in.
+
+                    // We must have at least one request asking for full semantics.
+                    Contract.ThrowIfFalse(changes.Any(c => !c.FrozenPartialSemantics));
+
+                    // All those requests should have cancellation tokens provided with them.
+                    Contract.ThrowIfFalse(changes.Where(c => !c.FrozenPartialSemantics).All(c => c.NonFrozenComputationToken != null));
+
+                    // Get the first non-cancelled token if present.
+                    var nonFrozenComputationToken = changes.FirstOrNull(t => t.NonFrozenComputationToken?.IsCancellationRequested is false)?.NonFrozenComputationToken;
```

Comment:
To me, this is misleading. There will be at most one non-cancelled token, and if there is a non-cancelled token, it will be the last one. Due to the previous assertions, this is literally the last item in the entire `changes` list, so you can use `^1` or equivalent.

Reply:
yup. done.